File: TESS_PythonAPI_demo/19自动驾驶/MyNet.py
# -*- coding: utf-8 -*-

import os
from pathlib import Path
import sys

from PySide2.QtCore import *
from PySide2.QtGui import *
from PySide2.QtWidgets import *

from Tessng import PyCustomerNet, TessInterface, TessPlugin, NetInterface, tngPlugin, tngIFace, m2p
from Tessng import NetItemType, GraphicsItemPropName

# 用户插件子类，代表用户自定义与路网相关的实现逻辑，继承自MyCustomerNet
class MyNet(PyCustomerNet):
    def __init__(self):
        super(MyNet, self).__init__()

    # 创建路网
    def createNet(self):
        # 代表TESS NG的接口
        iface = tngIFace()
        # 代表TESS NG的路网子接口
        netiface = iface.netInterface()

        # 第一条路段
        startPoint = QPointF(m2p(-300), 0)
        endPoint = QPointF(m2p(300), 0)
        lPoint = [startPoint, endPoint]
        link1 = netiface.createLink(lPoint, 2, "公路1")
        if link1 is not None:
            # 车道列表
            lanes = link1.lanes()
            # 打印该路段所有车道ID列表
            print("公路1车道ID列表：", [lane.id() for lane in lanes])
            # 在当前路段创建发车点
            dp = netiface.createDispatchPoint(link1)
            if dp != None :
                # 设置发车间隔，含车型组成、时间间隔、发车数
                dp.addDispatchInterval(1, 14, 9)



    # 过载的父类方法，当打开网后TESS NG调用此方法
    #     实现的逻辑是：路网加载后获取路段数，如果路网数为0则调用方法createNet构建路网，之后再次获取路段数，如果大于0则启动仿真
    def afterLoadNet(self):
        # 代表TESS NG的接口
        iface = tngIFace()
        # 代表TESS NG的路网子接口
        netiface = iface.netInterface()
        # 获取路段数
        count = netiface.linkCount()
        if(count == 0):
            self.createNet()
        if(netiface.linkCount() > 0):
            #所有路段
            lLink = netiface.links()
            #ID等于1的路段
            link = netiface.findLink(1)
            if link is not None:
                #路段中心线断点集
                lPoint = link.centerBreakPoints()
                print("一条路段中心线断点：", [(p.x(), p.y()) for p in lPoint])
                lLane = link.lanes()
                if lLane is not None and len(lLane) > 0:
                    #第一条车道中心线断点
                    lPoint = lLane[0].centerBreakPoints()
                    print("一条车道中心线断点：", [(p.x(), p.y()) for p in lPoint])
            #所有连接段
            lConnector = netiface.connectors()
            if lConnector is not None and len(lConnector) > 0:
                #第一条连接段的所有“车道连接”
                lLaneConnector = lConnector[0].laneConnectors()
                #其中第一条“车道连接”
                laneConnector = lLaneConnector[0]
                #"车道连接“断点集
                lPoint = laneConnector.centerBreakPoints()
                print("一条'车道连接'中心线断点：", [(p.x(), p.y()) for p in lPoint])

            #启动仿真
            iface.simuInterface().startSimu()

        # 下面注释掉的代码逻辑是：通过插件获取传入的配置对象config，从中获取属性'__simuafterload'值，如果等于True值启动仿真
        # plugin = tngPlugin()
        # config = plugin.tessngConfig()
        # if config['__simuafterload'] is True:
        #     iface.simuInterface().startSimu()


    ''' 过载的父类方法，在绘制路网元素时被调用，确定用ID或名称，以及字体大小绘制标签，也可确定不绘制标签
    itemType：NetItemType常量，代表不同类型路网元素
    itemId：路网元素的ID
    ref_outPropName：返回值，GraphicsItemPropName枚举类型，影响路段和连接段的标签是否被绘制
        GraphicsItemPropName。None_表示不绘制，GraphicsItemPropName。ID：表示绘制ID，GraphicsItemPropName。NAME:表示绘制名称
    ref_outFontSize，返回值，标签大小，单位：米。假设车道宽度是3.5米，如果ref_outFontSize.value等于7，绘制的标签大小占两个车道宽度
    '''
    def ref_labelNameAndFont(self, itemType, itemId, ref_outPropName, ref_outFontSize):
        # 代表TESS NG的接口
        iface = tngIFace()
        # 代表TESS NG仿真子接口
        simuiface = iface.simuInterface()
        # 如果仿真正在进行，设置ref_outPropName.value等于GraphicsItemPropName.None_，路段和车道都不绘制标签
        if simuiface.isRunning():
            ref_outPropName.value = GraphicsItemPropName.None_
            return
        # 默认绘制ID
        ref_outPropName.value = GraphicsItemPropName.Id
        # 标签大小为6米
        ref_outFontSize.value = 6
        # 如果是连接段一律绘制名称
        if itemType == NetItemType.GConnectorType:
            ref_outPropName.value = GraphicsItemPropName.Name
        elif itemType == NetItemType.GLinkType:
            if itemId == 1 or itemId == 5 or itemId == 6:
                ref_outPropName.value = GraphicsItemPropName.Name

    # 过载父类方法，是否绘制车道中心线
    def isDrawLaneCenterLine(self, laneId):
        return True

    # 过载父类方法，是否绘制路段中心线
    def isDrawLinkCenterLine(self, linkId):
        if linkId == 1:
            return False
        else:
            return True

File: TESS_CppAPI_EXAMPLE/MyNet.cpp
#include "MyNet.h"

#include <QMessageBox>

#include "IRoadNet.h"
#include "tessinterface.h"
#include "netinterface.h"
#include "simuinterface.h"
#include "Plugin/_netitemtype.h"
#include "IDispatchPoint.h"
#include "ilink.h"
#include "IConnector.h"
#include "ILaneConnector.h"
#include "UnitChange.h"

MyNet::MyNet(){
}

//加载路网前的准备
void MyNet::beforeLoadNet() {
	//QMessageBox::information(nullptr, QString(), QString("加载路网前的准备"));
}

//加载完路网后的行为
void MyNet::afterLoadNet() {
	if (gpTessInterface->netInterface()->linkCount() == 0) {
		//软件打开后自动加载临时路网，因为临时路网上没有路段，所以下面创建路段等要素的语句得到执行
		createNet();
	}



	//2022-07-18
	//获取所有路段
	QList<ILink*> lLink =  gpTessInterface->netInterface()->links();
	int minPointCount = 10000;
	foreach(ILink * pLink, lLink)
	{
		//==============创建当前路段SVG=================
		//路段中心线断点集
		QList<QPointF> lPoints = pLink->centerBreakPoints();
		//路段左侧断点集
		QList<QPointF> lLeftPoints = pLink->leftBreakPoints();
		//路段右侧断点集
		QList<QPointF> lRightPoints = pLink->rightBreakPoints();
		//创建路段SVG,略

		//===============创建当前路段所有车道SVG================
		//路段的所有车道
		QList<ILane*> lLane = pLink->lanes();
		foreach(ILane * pLane, lLane)
		{
			if (pLane->centerBreakPoint3Ds().size() < minPointCount)
			{
				minPointCount = pLane->centerBreakPoint3Ds().size();
			}
			//车道中心线断点集
			QList<QPointF> lPoint = pLane->centerBreakPoints();
			//车道左侧断点集
			QList<QPointF> lLeftPoint = pLane->leftBreakPoints();
			//车道右侧断点集
			QList<QPointF> lRightPoint = pLane->rightBreakPoints();
		}
	}
	qDebug() << minPointCount;

	//2022-07-21
	//获取所有连接段
	QList<IConnector*> lConnector = gpTessInterface->netInterface()->connectors();
	foreach(IConnector* pConn, lConnector)
	{
		QList<ILaneConnector*> lLC = pConn->laneConnectors();
		foreach(ILaneConnector * pLC, lLC)
		{
			//"车道连接”左侧断点集
			QList<QPointF> lLeftPoint = pLC->leftBreakPoints();
			//“车道连接”右侧点集
			QList<QPointF> lRightPoint = pLC->rightBreakPoints();
			//=========编制svg===========

		}
	}

	//如果配置文件配置了加载后启动仿真，则TESSNG会启动仿真，本方法不用启动仿真，否则本方法启动仿真
	QJsonObject config =  gpTessPlugin->tessngConfig();
	if (config.value("__simuafterload").isBool() && config.value("__simuafterload").toBool())
	{
		return;
	}

}

//创建路网
void MyNet::createNet() {
	IRoadNet *pRoadNet = gpTessInterface->netInterface()->setNetAttrs("API创建路网", "OPENDRIVE");
	qDebug() << pRoadNet->netName();

	//第一条路段
	QPointF startPoint = QPointF(m2p(-300), 0);
	QPointF endPoint = QPointF(m2p(300), 0);
	QList<QPointF> lPoint;
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink1 = gpTessInterface->netInterface()->createLink(lPoint, 7, "曹安公路");
	if (pLink1) {
		//创建发车点
		IDispatchPoint *pDp = gpTessInterface->netInterface()->createDispatchPoint(pLink1);
		if (pDp) {
			pDp->addDispatchInterval(1, 2, 28);
		}
	}

	//创建第二条路段
	startPoint = QPointF(m2p(-300), m2p(-25));
	endPoint = QPointF(m2p(300), m2p(-25));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink2 = gpTessInterface->netInterface()->createLink(lPoint, 3, "次干道");
	if (pLink2) {
		//创建发车点
		IDispatchPoint *pDp = gpTessInterface->netInterface()->createDispatchPoint(pLink2);
		if (pDp) {
			pDp->addDispatchInterval(1, 3600, 3600);
		}
		//修改车道类型
		pLink2->lanes()[0]->setLaneType(QObject::tr("公交专用道"));
	}

	//创建第三条路段
	startPoint = QPointF(m2p(-300), m2p(25));
	endPoint = QPointF(m2p(-150), m2p(25));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink3 = gpTessInterface->netInterface()->createLink(lPoint, 3);
	if (pLink3) {
		//创建发车点
		IDispatchPoint *pDp = gpTessInterface->netInterface()->createDispatchPoint(pLink3);
		if (pDp) {
			pDp->addDispatchInterval(1, 3600, 3600);
		}
	}

	//创建第四条路段
	startPoint = QPointF(m2p(-50), m2p(25));
	endPoint = QPointF(m2p(50), m2p(25));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink4 = gpTessInterface->netInterface()->createLink(lPoint, 3);

	//创建第五条路段
	startPoint = QPointF(m2p(150), m2p(25));
	endPoint = QPointF(m2p(300), m2p(25));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink5 = gpTessInterface->netInterface()->createLink(lPoint, 3, "自定义限速路段");
	if (pLink5) {
		pLink5->setLimitSpeed(30);
	}

	//创建第六条路段
	startPoint = QPointF(m2p(-300), m2p(50));
	endPoint = QPointF(m2p(300), m2p(50));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink6 = gpTessInterface->netInterface()->createLink(lPoint, 3, "动态发车路段");
	if (pLink6) {
		pLink6->setLimitSpeed(80);
	}

	//创建第七条路段
	startPoint = QPointF(m2p(-300), m2p(75));
	endPoint = QPointF(m2p(-250), m2p(75));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink7 = gpTessInterface->netInterface()->createLink(lPoint, 3);
	if (pLink7) {
		pLink7->setLimitSpeed(80);
	}

	//创建第八条路段
	startPoint = QPointF(m2p(-50), m2p(75));
	endPoint = QPointF(m2p(300), m2p(75));
	lPoint.clear();
	lPoint << startPoint;
	lPoint << endPoint;
	ILink *pLink8 = gpTessInterface->netInterface()->createLink(lPoint, 3);
	if (pLink8) {
		pLink8->setLimitSpeed(80);
	}

	//创建第一条连接段
	if (pLink3 && pLink4) {
		QList<int> lFromLaneNumber = QList<int>() << 1 << 2 << 3;
		QList<int> lToLaneNumber = QList<int>() << 1 << 2 << 3;
		IConnector *pConn1 = gpTessInterface->netInterface()->createConnector(pLink3->id(), pLink4->id(), lFromLaneNumber, lToLaneNumber
			, "连接段1");
	}

	//创建第二条连接段
	if (pLink4 && pLink5) {
		IConnector *pConn2 = gpTessInterface->netInterface()->createConnector(pLink4->id(), pLink5->id(), QList<int>() << 1 << 2 << 3, QList<int>() << 1 << 2 << 3, "连接段2");
	}

	//创建第三条连接段
	if (pLink7 && pLink8) {
		IConnector *pConn3 = gpTessInterface->netInterface()->createConnector(pLink7->id(), pLink8->id(), QList<int>() << 1 << 2 << 3, QList<int>() << 1 << 2 << 3, "动态发车连接段");
	}

	//关闭主窗体或重新打开新的路网，会从内存删除当前路网元素，不必手动删除。
}

//写标签，按照给定的属性名和字体大小（米）
void MyNet::labelNameAndFont(int itemType, long itemId, int &outPropName, qreal &outFontSize) {
	outPropName = GraphicsItemPropName::Id;
	outFontSize = 6;
	if (itemType == NetItemType::GConnectorType) {
		outPropName = GraphicsItemPropName::Name;
	}
	else if (itemType == NetItemType::GLinkType) {
		if (itemId == 1 || itemId == 5 || itemId == 6) {
			outPropName = GraphicsItemPropName::Name;
		}
	}
}

MyNet::~MyNet(){
}